### ISSUE TITLE
[feature-request] Support for Managed NAT Gateway

### ISSUE THREAD

**Issue by cjs** [hashicorp/terraform#4374]
https://aws.amazon.com/blogs/aws/new-managed-nat-network-address-translation-gateway-for-aws/

This just came out, but wanted to get it into the backlog.

**Comment by jen20** [comment]
:+1: We'll do this as soon as support lands in the AWS SDK for Go!

**Comment by nrcxcia** [comment]
+1

**Comment by conorgil** [comment]
+1

This would save SO much time and effort! Very excited to see this released by AWS and excited for support in Terraform. The announcement blog post also links to [more docs](http://docs.aws.amazon.com/AmazonVPC/latest/UserGuide/vpc-nat-gateway.html) .

I think supporting this will require the following updates:

1. Create a new resource `aws_nat_gateway`, which supports something like the following to create a NAT Gateway for each AZ:
    
    ```hcl
    resource "aws_nat_gateway" "main" {
        count = "${length(split(",", var.zones))}"

        public_subnet = "${aws_subnet.public.*.id}"
        elastic_ip_address = "${aws_eip.nat.*.id}"
        tags {
          Name = "${var.environment_name}_ngw_for_${element(split(",", var.zones), count.index)_az}"
          environment_name = "${var.environment_name}"
        }
      }
    ```
2. Update [aws_route_table](https://www.terraform.io/docs/providers/aws/r/route_table.html) and [aws_route](https://www.terraform.io/docs/providers/aws/r/route.html) to support the new NAT gateway. Since those resources already support the `gateway_id` argument, that could be overloaded to be either the IGW or new NGW resource. If that is complicated to implement, then we could create a new argument `nat_gateway_id` or something like that.

I am happy to help with docs or anything else non-programming. I haven't dove into Go yet and don't have time at the moment. Let me know how I can help.

**Comment by gamename** [comment]
+1

**Comment by puredistortion** [comment]
:+1:  We are excited to integrate this into our standard deployment

**Comment by majormoses** [comment]
+1

**Comment by mtb-xt** [comment]
AWS SDK for Go has just released version 1.0.6, it supports NAT Gateways

**Comment by jen20** [comment]
This is started in #4381.

**Comment by timothykimball** [comment]
+1
Whoop!

**Comment by jen20** [comment]
Landed in #4381!

**Comment by rajholla** [comment]
Do we have "data source" support for nat gateway's yet? 
This is very useful when we need to fetch the  NAT gateay id's from existing VPC.

**Comment by majormoses** [comment]
@rajholla there is no data source for it yet but you could import your existing gateways: https://www.terraform.io/docs/providers/aws/r/nat_gateway.html#import

**Comment by nmarchini** [comment]
Is there going to be support for Tags, the current doumentation doesn't allow this and I'm getting the following error when trying to add tags

Errors:
  * aws_nat_gateway.core_nat: : invalid or unknown key: tags

**Comment by bflad** [comment]
@nmarchini Terraform 0.10.0 now has split providers, like AWS, from the core Terraform repository into their own provider repositories. Please see https://github.com/terraform-providers/terraform-provider-aws/pull/1625 which is merged into master of terraform-provider-aws and should be released with version 1.0.0 of the provider.
